solr/solrj/src/java/org/apache/solr/common/assign/ReplicaAssignStrategy.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.common.assign;
+
+
+import org.apache.solr.client.solrj.cloud.SolrCloudManager;
+import org.apache.solr.client.solrj.cloud.autoscaling.Suggester;
+import org.apache.solr.common.MapWriter;
+import org.apache.solr.common.cloud.Replica;
+import org.apache.solr.common.params.CollectionParams;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The implementation class can be stored in clusterprops.json as follows
+ * {
+ *     assign-strategy : {
+ *         class : class-name
+ *         package-version: package-version
+ *     }
+ * }
+ * A user can load an assign-strategy from a package stored in package store and load it using the prefix notation
+ * e.g: assign-strategy: packag-ename:fully.qualified.ClassName. If the package name prefix is used , ensure that package-version
+ * is provided.
+ * If this is configured, Solr will use this everywhere replicas need to be assigned
+ *
+ * Every implementation of assign-strategy should implemenmt this interface
+ * A new instance is created just in time for computing operations. This instance will
+ * be thrown away after the computations are over
+ *
+ */
+public interface ReplicaAssignStrategy {
+
+    /**
+     * This method is invoked before anything else is done.
+     * choose to read any configuration that may be stored in ZK etc.
+     */
+    void init(SolrCloudManager cloudManager);
+
+    /**
+     * get a list of operations that can be performed for the intents.
```

Comment:
"can" or "will"? This is an important difference in semantics, because if you know you will perform these ops then you should make reservations for the positions. "Can" is much weaker, it doesn't lock out the slots for other concurrent computations.

Reply:
We do not know if the invoker "will" execute these  operations.

Reply:
In other words, the assumption that this method is using the current cluster state that was provided to it and itself it does not make any reservations.

(Just clarifying ... I think that's ok, we could provide it a cluster view that already includes other operations in progress).